src/core/lv2/Lv2Proc.cpp

```diff
@@ -0,0 +1,545 @@
+										nullptr, dispName));
+					break;
+				case Lv2Ports::Vis::Enumeration:
+				{
+					ComboBoxModel* comboModel
+						= new ComboBoxModel(
+							nullptr, dispName);
+					const LilvScalePoints* sps =
+						lilv_port_get_scale_points(m_plugin, lilvPort);
+					LILV_FOREACH(scale_points, i, sps)
+					{
+						const LilvScalePoint* sp = lilv_scale_points_get(sps, i);
+						ctrl->m_scalePointMap.push_back(lilv_node_as_float(
+										lilv_scale_point_get_value(sp)));
+						comboModel->addItem(
+							lilv_node_as_string(
+								lilv_scale_point_get_label(sp)));
+					}
+					ctrl->m_connectedModel.reset(comboModel);
+					break;
+				}
+				case Lv2Ports::Vis::Toggled:
+					ctrl->m_connectedModel.reset(
+						new BoolModel(static_cast<bool>(meta.m_def),
+										nullptr, dispName));
+					break;
+			}
+		}
+		port = ctrl;
+	}
+	else if (meta.m_type == Lv2Ports::Type::Audio)
+	{
+		Lv2Ports::Audio* audio =
+			new Lv2Ports::Audio(
+					static_cast<std::size_t>(
+						Engine::mixer()->framesPerPeriod()),
+					portIsSideChain(m_plugin, lilvPort)
+				);
+		port = audio;
+	} else {
+		port = new Lv2Ports::Unknown;
+	}
+
+	*dynamic_cast<Lv2Ports::Meta*>(port) = meta;
+	port->m_port = lilvPort;
+	port->m_plugin = m_plugin;
+}
+
+
+
+
+void Lv2Proc::createPorts()
+{
+	// register ports at the processor after creation,
+	// i.e. link their data or count them
+	struct RegisterPort : public Lv2Ports::Visitor
+	{
+		Lv2Proc* proc;
+
+		void visit(Lv2Ports::Control& ctrl) override
+		{
+			if(ctrl.m_flow == Lv2Ports::Flow::Input)
+			{
+				AutomatableModel* amo = ctrl.m_connectedModel.get();
+				proc->m_connectedModels.emplace(
+					lilv_node_as_string(lilv_port_get_symbol(
+										proc->m_plugin, ctrl.m_port)),
+					amo);
+				proc->addModel(amo);
+
+				++proc->m_controlCount;
+			}
+		}
+
+		void visit(Lv2Ports::Audio& audio) override
+		{
+			if(!audio.isSideChain())
+			{
+				StereoPortRef dummy;
+				StereoPortRef* portRef = &dummy;
+				switch(audio.m_flow)
+				{
+					case Lv2Ports::Flow::Input:
+						portRef = &proc->m_inPorts;
+						break;
+					case Lv2Ports::Flow::Output:
+						portRef = &proc->m_outPorts;
+						break;
+					case Lv2Ports::Flow::Unknown:
+						break;
+				}
+				// in Lv2, leftPort is defined to be the first port
+				if(!portRef->m_left) {
+					portRef->m_left = &audio;
+				} else if(!portRef->m_right) {
+					portRef->m_right = &audio;
+				}
+			}
+		}
+	};
+
+	unsigned maxPorts = lilv_plugin_get_num_ports(m_plugin);
+	m_ports.resize(maxPorts);
+
+	for(unsigned portNum = 0; portNum < maxPorts; ++portNum)
+	{
+		createPort(portNum);
+		RegisterPort registerPort;
+		registerPort.proc = this;
+		m_ports[portNum]->accept(registerPort);
+	}
+
+	// initially assign model values to port values
+	copyModelsFromCore();
+
+	// debugging:
+	//dumpPorts();
+}
+
+
+
+
+struct ConnectPorts : public Lv2Ports::Visitor
+{
+	unsigned m_num;
+	LilvInstance* m_instance;
+	void con(void* location) {
```

Comment:
Space here as well since syntax looks like if/else/etc ?

Reply:
It's a function definition, those are specified without space in the conventions. So I'd leave this as it is.